data_analysis/gtfs_fetcher.py

```diff
@@ -0,0 +1,117 @@
+    feed_start_date: str
+    feed_end_date: str
+    schedule_source: ScheduleSource
+
+    def __str__(self):
+        if self.schedule_source == ScheduleSource.TRANSITFEEDS:
+            label = ''
+        else:
+            label = '_cta'
+        return f'v_{self.schedule_version}_fs_{self.feed_start_date}_fe_{self.feed_end_date}{label}'
+
+    def __getitem__(self, item):
+        if item not in frozenset(['schedule_version', 'feed_start_date', 'feed_end_date']):
+            raise KeyError(item)
+        return self.__dict__[item]
+
+    @classmethod
+    def from_pendulum(cls, version, start_date, end_date, schedule_source: ScheduleSource):
+        return cls(version.format("YYYYMMDD"),
+                   start_date.format("YYYY-MM-DD"),
+                   end_date.format("YYYY-MM-DD"),
+                   schedule_source)
+
+    @classmethod
+    def from_dict(cls, d):
+        return cls(d['schedule_version'],
+                   d['feed_start_date'],
+                   d['feed_end_date'],
+                   ScheduleSource.TRANSITFEEDS)
+
+    def interval(self):
+        start = pendulum.parse(self.feed_start_date)
+        end = pendulum.parse(self.feed_end_date)
+        return pendulum.interval(start, end)
+
+    def contains(self, date_str: str) -> bool:
+        d = pendulum.parse(date_str)
+        return d in self.interval()
+
+
+class GTFSFetcher:
+    def __init__(self, cache_manager: CacheManager):
+        self.cache_manager = cache_manager
+        files = s3.list_objects_v2(Bucket=BUCKET_PUBLIC, Prefix='cta_schedule_zipfiles_raw/')
+        self.unique_files = {}
+        self.versions = {}
+        for fc in files['Contents']:
+            key = fc['ETag']
+            filename = fc['Key'].split('/')[1]
+            if not filename.startswith('google_transit_'):
+                continue
+            size = fc['Size']
+            version = filename.removeprefix('google_transit_').removesuffix('.zip').replace('-', '')
+            self.unique_files.setdefault(key, []).append((filename, size, fc['Key'], version))
+        for v in self.unique_files.values():
+            v.sort()
+            tup = v[0]
+            self.versions[tup[-1]] = tup
+
+    def list(self):
+        tups = []
+        for v in self.unique_files.values():
+            tups.append(v[0])
+        tups.sort()
+        return tups
+
+    def get_versions(self):
+        return list(sorted(self.versions.keys()))
+
+    def retrieve_file(self, schedule_feed_info: ScheduleFeedInfo):
+        version_id = schedule_feed_info.schedule_version
+        local_filename, _, s3_filename, _ = self.versions[version_id]
+        if schedule_feed_info.schedule_source == ScheduleSource.S3:
+            url = f'https://{BUCKET_PUBLIC}.s3.us-east-2.amazonaws.com/{s3_filename}'
+            return self.cache_manager.retrieve('cta_zipfiles', local_filename, url)
+        elif schedule_feed_info.schedule_source == ScheduleSource.TRANSITFEEDS:
```

Comment:
Yes, good idea. I didn't realize that we had already scraped all of the relevant transitfeeds schedules.

Note that reading from transitfeeds is only necessary at this point if you're rebuilding the entire history. The incremental workflow doesn't need it. Since this PR is already large, would it make sense to refactor the transitfeeds stuff as a followup?

Reply:
Split to new issue: https://github.com/chihacknight/chn-ghost-buses/issues/85